### PR TITLE
Add markdown toolbar to V2 editor

### PR DESCRIPTION
**Pull request by aitchiss**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR adds the markdown toolbar to the V2 editor 🎉

It also changes the image upload functionality in the V2 editor, described in the QA steps below. The image upload process for V1 editor remains the same (we might wish to change that in future), and the native iOS functionality should be unchanged, other than you now press the toolbar image button to begin selecting a picture.

A large proportion of the markdown toolbar functionality has been previously tested in Storybook, so the QA steps below focus primarily on the integration with the editor, and the changes to the image uploader.

## Related Tickets & Documents

Epic for toolbar: https://github.com/forem/forem/issues/14808
Closes https://github.com/forem/forem/issues/15115

There is [another small change in progress](https://github.com/forem/forem/issues/15294) which updates the logic for inserting new lines on either side of a block syntax (e.g. quote, code block). That work isn't merged yet, but when it is, I'll have some tests to update here (marked with TODOs).

## QA Instructions, Screenshots, Recordings

Testing for this feature should cover:

1. Integration of toolbar with editor text area
2. Image uploader on non-iOS devices
3. Image uploader on native iOS device
4. General usability/concept check on a range of browsers, screen sizes and devices
5. Regression test of V1 editor image upload

### 1. Integration of toolbar with editor text area

- Run the app locally and navigate to `/new`
- Experiment with using the markdown toolbar formatter buttons
- Experiment with the keyboard shortcuts (displayed in a tooltip where available for: bold, italic, link, underline, strikethrough)
- Check that you can still `@mention` users as before
- Check that you can still drag and drop images into the text area as before
- Check that you can still copy/paste images in the text area as before

#### Happy path

- Run the app locally and navigate to `/new`
- Click the image button in the toolbar and pick an image to upload
- As soon as an image is selected you should see a spinner over the image button, and `Uploading...` in the correct spot in the editor
- Once the image successfully uploads you should see the `Uploading...` text replaced with `Image description`

https://user-images.githubusercontent.com/20773163/141308480-34c6c045-07a3-426e-b433-bc97aaa0161e.mp4

#### User cancels request

- In your browser dev tools, throttle your network connection so that you have time to complete the next test
- Press the image button and choose a file to upload
- Hover over the spinner on the image button - it should change to an 'X' and have the tooltip 'Cancel upload'
- Click the button to cancel the upload
- The placeholder `Uploading...` should disappear and a snackbar should be shown noting that you aborted the request

https://user-images.githubusercontent.com/20773163/141308283-0ebbf63d-e6bb-495a-b38e-7cc7abd5817a.mp4

#### Error path

- Remove the throttle in your dev tools, and instead block the network request to `/image_uploads`
- Click the image button and select an image
- Check when the request fails, the placeholder `Uploading...` is removed, and a snackbar is shown

https://user-images.githubusercontent.com/20773163/141308173-cd665345-2974-418b-a72e-b06873202dea.mp4

### 3. Image uploader on native iOS device

- Image uploader on an iOS device should use the native image picker, but in all other ways it should behave the same as above
- If you have an iOS device to hand you can use that to test (I'm hoping someone will!), or if you have XCode/simulator installed you can run `open -a Simulator.app` in the terminal to open up a simulator to test on by opening up the browser and navigating to localhost

https://user-images.githubusercontent.com/20773163/141311870-741ffb16-8499-4916-b9fe-7c760d9ac56c.mp4

### 4. General usability/concept check on a range of browsers, screen sizes and devices

- UI should look consistent across browsers and devices, including mobile devices

### 5. Regression test of V1 editor image upload

- The V1 editor image picker should function exactly as before, and the toolbar should not be shown

### UI accessibility concerns?

- Toolbar accessibility concerns have mostly been addressed in preceding PRs to add the toolbar to storybook
- Image upload success is communicated to screen reader users via an `aria-live`. We didn't have this in place before, but given a user may crack on with typing their post while the image is uploading, I thought it made sense to provide this notification
- Image upload failure or cancellation is communicated to screen reader users via an alert, as part of our snackbar component
- The new image uploader makes use of a visually hidden image file input, which is triggered by the toolbar button. The file input is only focusable either by JS, so doesn't interrupt the tab order, but screen reader users who pull up a list of all form controls will become aware of two - seemingly identical - means to upload an image. On balance, I think this is OK as both methods function, and the vast majority of users won't be aware at all of that hidden file input.

## Added/updated tests?

- [X] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [X] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

Liz Lemon says 'let's do this'

### CODE REVIEW

File: app/assets/stylesheets/components/editor-toolbar.scss
Comment (aitchiss) [inline app/assets/stylesheets/components/editor-toolbar.scss:5]: This new css controls the image upload "cancel" button - which displays a spinner unless you hover or focus the button (in which case you see the cancel icon)

I looked at tailwind-y ways to achieve this but I don't think we currently use e.g. `hover:opacity-0`

File: app/javascript/.storybook/preview-head.html
Comment (aitchiss) [inline app/javascript/.storybook/preview-head.html:6]: as this check happens in the ImageUploader, we need to stub this in storybook so we can continue to show the toolbar component there

File: app/javascript/article-form/actions.js
Comment (aitchiss) [inline app/javascript/article-form/actions.js:86]: `signal` allows us to pass a means to cancel the request

File: app/javascript/crayons/MarkdownToolbar/MarkdownToolbar.jsx
Comment (aitchiss) [inline app/javascript/crayons/MarkdownToolbar/MarkdownToolbar.jsx:14]: Previously we managed the roving tab index by going directly to next or previous element siblings. Now we introduce the image uploader we can't do that, and have to instead check for a selector match (because there is an additional - hidden - file input in the ImageUploader that we don't want to focus)

**Comment by aitchiss** [comment]
@fdoxyz added you on review in light of the refactors around the image upload - I've tested out the image picker using the simulator and all looks OK to me, but would definitely appreciate a quick look to make sure I'm not breaking anything 😄 🙏

File: app/javascript/article-form/components/ArticleCoverImage.jsx
Comment (jeremyf) [inline app/javascript/article-form/components/ArticleCoverImage.jsx:83]: Definitely like naming these parameters!

**Review by fdoxyz — APPROVED** [review]
This is really cool! Everything worked as expected on Desktop, Android (web) and iOS (web + app) 🙌 

I have a [refactor PR](https://github.com/forem/forem/pull/15212) for how the native integration works, but it seems like it shouldn't be too tricky if either one gets merged first since they work on separate functions/components (I have a feeling this one might go in first haha). Excited about this! 

A tiny **non-blocking nitpick/suggestion**: Maybe we can get rid of the horizontal scroll indicator? I don't see it on iOS but Chrome on Android does look this way:

Screenshot_20211112-080156_Chrome

If we want to ensure users know there's a horizontal scroll (_with a more native feel_) it may be worth adding a gradient to the very far end that makes it seem like the last element "half hides behind it", so an intuitive user would scroll to uncover what may be available in the far right end. 

More than okay to tackle this tiny enhancement in a separate PR though if we rather to discuss this further 🙂

**Comment by fdoxyz** [comment]
> but I guess whatever we do there might be some screen sizes where it's not super obvious it's scrollable 🤔

Going for a real fancy/detailed result gets tricky indeed, specially because of the variable size screens. I guess this would only affect screen widths that are under size M or something along those lines, where we know there will be a horizontal overflow.

In a native apps one thing I've seen in the past is that elements in the horizontal scroll _never match a multiple of the screen width_. That way the elements are always "cut in half", making it super obvious you can scroll. Also other design details (that are way above what I could ever come up with) like "giving the container a depth" might help.

Just spitballing ideas, they may or may not be feasible/necessary 😄

**Comment by aitchiss** [comment]
I've hidden the scrollbar on the smaller layout, and I've removed the right-hand padding on the parent container in that case, as I noticed it was hiding a bit of the icons at the right. I think in most cases, it suggests a scroll because either an icon is cut off, or there isn't the same amount of padding visible after the icon... it's quite subtle but I guess we only need people to discover it once 😄 

Chrome Pixel 2 emulator showing the layout described above

File: app/javascript/article-form/components/Toolbar.jsx
Comment (aitchiss) [reply app/javascript/article-form/components/Toolbar.jsx:9]: Yes it's a small thing but when `version === 'v1'` evaluates to false, we end up adding a className "false". It's harmless, but unnecessary and potentially confusing. Replacing with the ternary means we just skip adding the className if `false`, which I'm 99.99% sure is the original intention

File: app/javascript/crayons/MarkdownToolbar/MarkdownToolbar.jsx
Comment (aitchiss) [reply app/javascript/crayons/MarkdownToolbar/MarkdownToolbar.jsx:14]: Yes, good point, I'll add that in

**Review by jeremyf — APPROVED** [review]
Though please wait to merge until Monday!

File: app/assets/stylesheets/views/article-form.scss
Comment (nickytonline) [inline app/assets/stylesheets/views/article-form.scss:180]: For the toolbar, `&__toolbar`, I imagine we can use a utility class like `m:p-0` instead on the tag?

File: app/javascript/article-form/actions.js
Comment (nickytonline) [inline app/javascript/article-form/actions.js:95]: Refactor (blocking): Since this is being updated, use the `@utiltities/http/request` utility function,
```suggestion
  request('/image_uploads', {
    method: 'POST',
    body: generateUploadFormdata(payload),
    signal,
  })
```

File: app/javascript/crayons/MarkdownToolbar/MarkdownToolbar.jsx
Comment (nickytonline) [inline app/javascript/crayons/MarkdownToolbar/MarkdownToolbar.jsx:187]: Refactor (blocking): Localize the placeholder text. It's used in at least two places in the code.

File: app/javascript/crayons/MarkdownToolbar/__tests__/MarkdownToolbar.test.jsx
Comment (nickytonline) [inline app/javascript/crayons/MarkdownToolbar/__tests__/MarkdownToolbar.test.jsx:11]: This is out of scope of this PR, but we should formalize some mocks for certain things so they can be reused instead of copy pasta.

**Review by nickytonline — CHANGES_REQUESTED** [review]
This is looking great! A couple change requests.

File: app/javascript/crayons/MarkdownToolbar/MarkdownToolbar.jsx
Comment (aitchiss) [reply app/javascript/crayons/MarkdownToolbar/MarkdownToolbar.jsx:187]: If this is a blocker I can add it, but it would require a change to what translations we load into the frontend (currently we only include those in the `core` namespace, but this would live in `views.editor`:

https://github.com/forem/forem/blob/09828853f6cc0e07da74f4da885dfa93f5204662/app/views/layouts/application.html.erb#L111

I feel like doing that should be a separate PR - what do you think?

If we left that for later i18n work, I could just consolidate the couple of occurrences of this string into a constant for now

File: app/javascript/article-form/actions.js
Comment (aitchiss) [reply app/javascript/article-form/actions.js:95]: The only thing is that the `request` function assumes any request body should be a JSON string, whereas this needs to be a `FormData` object - I feel like it may be the exception to using the standard `request` helper. What do you think?

Otherwise we'll need to add some refactors to the helper that would allow a body to be either FormData or JSON string

File: app/assets/stylesheets/views/article-form.scss
Comment (aitchiss) [reply app/assets/stylesheets/views/article-form.scss:180]: I had a look at this, but it's tricky because those responsive helpers are designed to take effect from the given breakpoint, but I want to target anything _until_ that breakpoint (e.g. `m:p-0` would do the opposite of what I want to do).

The padding for this toolbar is already declared in this file so we can use the `--content-padding` variable, so it seemed best to keep this styling together.

File: app/javascript/crayons/MarkdownToolbar/__tests__/MarkdownToolbar.test.jsx
Comment (aitchiss) [reply app/javascript/crayons/MarkdownToolbar/__tests__/MarkdownToolbar.test.jsx:11]: Yeah I think that would be a good plan 👍

File: app/assets/stylesheets/views/article-form.scss
Comment (ludwiczakpawel) [reply app/assets/stylesheets/views/article-form.scss:180]: How about this... May seem over-engineering at first glance, but I think it's worth for keeping min-widths over max-widths for our breakpoints.. Thoughts? (also, I haven't tested it :D)

```
&__toolbar {
  padding: var(--su-2) var(--content-padding-x);
  padding-right: var(--toolbar-padding-right, 0);

  @media (min-width: $breakpoint-m) {
    --toolbar-padding-right: var(--content-padding-x);
  }
}
```

Comment (aitchiss) [reply app/assets/stylesheets/views/article-form.scss:180]: That sounds like a good plan 🙌

File: app/javascript/article-form/actions.js
Comment (nickytonline) [reply app/javascript/article-form/actions.js:95]: Yeah, makes sense. We can revisit modifying `request`.

**Review by nickytonline — APPROVED** [review]
This is great work @aitchiss! 🚀

**Comment by aitchiss** [comment]
Converted to draft to avoid any accidental merges while we plan the release comms for this one

**Comment by aitchiss** [comment]
Hey @fdoxyz I've tagged you for re-review on this one, as I've just updated following your changes in https://github.com/forem/forem/pull/15212

It all seems to be working for me, but would definitely appreciate another set of eyes on the mobile app integration 🙏

**Review by fdoxyz — APPROVED** [review]
**tl;dr** I gave it another run and things still work as expected 🎉 

There is one thing that may have changed in the `ImageUploader` though (maybe as it was refactored a bit), but it's _sort of unrelated_ with this PR. It sounds a bit weird 😅  but here's my attempt at elaborating on this:

- Right now the native image picker is disabled ([details here](https://github.com/forem/forem/pull/15212#discussion_r755507405)), so things work perfectly fine using the web implementation
   - When I say _web implementation_ I refer to it as using the default logic (as if it were a web browser without the native image picker integration)
- The problem I found was when re-enabling the native picker again. I believe [this EventListener](https://github.com/forem/forem/blob/main/app/javascript/article-form/components/ImageUploader.jsx#L219) is being registered multiple times
   - This causes one event (coming from the native picker) to paste multiple times the same image link
   - The reason it's registering multiple times I think comes from [this explanation](https://stackoverflow.com/a/47337711/3462026) (which is basically quoting [MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#Multiple_identical_event_listeners) 😄 )
      - _"Note, however that when using an anonymous function as the handler, such listeners will NOT be identical, because anonymous functions are not identical even if defined using the SAME unchanging source-code called repeatedly, even if in a loop."_
   - **Long story short:** I believe the reason this happens is that `render` is being called multiple times, but since the EventListener is being created with what seems like (my uneducated guess) an anonymous function, then one event coming in to inject the upload link is in turn adding multiple links in the Post editor

I could use your help @aitchiss to sort this out in #15487 in the coming days though. I'm just unsure of where/how in the `ImageUploader` component I can use a `removeEventListener` (I'm guessing there's a callback/hook for then the component gets de-registered or re-rendered).

Anyways, we don't really need to fix this right away in this PR, so this has my approval to go live as-is  ✅

**Comment by aitchiss** [comment]
Thanks for checking this, and great find @fdoxyz!

Definitely happy to help sorting this so we don't get multiple event listeners. I think your assessment is pretty much right - because the function `handleNativeMessage` is declared inside the component, it gets recreated every time, so gets treated as a new event listener on each iteration.

 My first (as yet untested thought) is that we can likely solve this by:

1. Moving `handleNativeMessage` out of the component
2. Attaching the listener inside of a `useEffect` that only runs once, e.g.

```jsx
useEffect(() => {
  document.addEventListener('ForemMobile', handleNativeMessage);

  return () => document.removeEventListener('ForemMobile', handleNativeMessage);
}, [])
```

The empty dependency array should stop us from attaching this more than once, and in any case if we move `handleNativeMessage` out of the component, it's no longer being created new every render cycle, so attaching it more than once _shouldn't_ have a negative effect anyway

**Pull request merged**
